### PR TITLE
Clean up Kafka batch writer

### PR DESCRIPTION
**Pull request by nathan-artie**
- Pass context as a parameter
- Add a generic batch iterator and use it instead of `NewBatch`

### CODE REVIEW

File: lib/iterator/iterator.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
+package iterator
+
+type batchIterator[T any] struct {
+	items []T
+	index int
+	n     int
```
Comment (Tang8330) [inline lib/iterator/iterator.go:6]: What's n?

Comment (nathan-artie) [reply lib/iterator/iterator.go:6]: size lol, I can rename.

File: lib/iterator/iterator.go [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+	}
+}
+
+func (i *batchIterator[T]) HasNext() bool {
+	return i.index < len(i.items)
+}
+
+func (i *batchIterator[T]) Next() []T {
+	if !i.HasNext() {
+		return make([]T, 0)
```
Comment (Tang8330) [inline lib/iterator/iterator.go:23]: return nil?

File: lib/iterator/iterator.go (hunk 1 above)
Comment (Tang8330) [reply lib/iterator/iterator.go:6]: Seems like it's the batch size, maybe call it size or step instead?

**Review by Tang8330 — APPROVED** [review]
lgtm % c

**Pull request merged**